### PR TITLE
SchemaView: format, lint autofix, add typing + docs

### PR DESCRIPTION
**Pull request by ialarmedalien**

### CODE REVIEW

File: linkml_runtime/index/object_index.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 from typing import Any, Union
 from collections.abc import Mapping, Iterator
 
-from linkml_runtime import SchemaView
+from linkml_runtime.utils.schemaview import SchemaView
```
Comment (ialarmedalien) [inline linkml_runtime/index/object_index.py:16]: use full path for import (this was mostly to make it easier for me to find places where weird stuff was being imported from SchemaView)

File: linkml_runtime/utils/schemaview.py [hunk 2]
```diff
@@ -1,37 +1,68 @@
+"""SchemaView, a virtual schema layered on top of a schema plus its import closure."""
+
+from __future__ import annotations
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:3]: allows nicer ways of expressing typing (i.e. `X | Y`, etc.)

File: linkml_runtime/utils/schemaview.py [hunk 3]
```diff
@@ -358,32 +422,30 @@ def all_schema(self, imports: bool = True) -> list[SchemaDefinition]:
...
-    def ordered(self, elements: ElementDict, ordered_by: Optional[OrderedBy] = None) -> ElementDict:
-        """
-        Order a dictionary of elements with some ordering method in :class:`.OrderedBy`
-        """
+    def ordered(self, elements: ElementDict, ordered_by: OrderedBy | None = None) -> ElementDict:
+        """Order a dictionary of elements with some ordering method in :class:`.OrderedBy`."""
         if ordered_by in (OrderedBy.LEXICAL, OrderedBy.LEXICAL.value):
             return self._order_lexically(elements)
-        elif ordered_by in (OrderedBy.RANK, OrderedBy.RANK.value):
+        if ordered_by in (OrderedBy.RANK, OrderedBy.RANK.value):
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:437]: no need for `else` or `elif` after a `return`

File: linkml_runtime/utils/schemaview.py [hunk 4]
```diff
@@ -418,21 +480,16 @@ def _order_rank(self, elements: ElementDict) -> ElementDict:
...
+        """Sort classes such that if C is a child of P then C appears after P in the list."""
         clist = list(elements.values())
         slist = []  # sorted
         can_add = False
         while len(clist) > 0:
             for i in range(len(clist)):
                 candidate = clist[i]
                 can_add = False
-                if candidate.is_a is None:
+                if candidate.is_a is None or candidate.is_a in [p.name for p in slist]:
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:491]: collapse / combine `if` conditions

File: linkml_runtime/utils/schemaview.py [hunk 5]
```diff
@@ -443,113 +500,118 @@ def _order_inheritance(self, elements: DefDict) -> DefDict:
...
+        """Retrieve all classes from a schema.
+
         :param ordered_by: an enumerated parameter that returns all the classes in the order specified.
         :param imports: include imports closure
         :return: all classes in schema view
         """
         classes = copy(self._get_dict(CLASSES, imports))
-        classes = self.ordered(classes, ordered_by=ordered_by)
-        return classes
+        return self.ordered(classes, ordered_by=ordered_by)
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:513]: return directly, no need to assign to a variable

File: linkml_runtime/utils/schemaview.py [hunk 6]
```diff
@@ -443,113 +500,118 @@ def _order_inheritance(self, elements: DefDict) -> DefDict:
...
         slots = copy(self._get_dict(SLOTS, imports))
         if attributes:
             for c in self.all_classes().values():
                 for aname, a in c.attributes.items():
                     if aname not in slots:
                         slots[aname] = a
 
-        slots = self.ordered(slots, ordered_by=ordered_by)
-        return slots
+        return self.ordered(slots, ordered_by=ordered_by)
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:543]: return directly

File: linkml_runtime/utils/schemaview.py [hunk 7]
```diff
@@ -443,113 +500,118 @@ def _order_inheritance(self, elements: DefDict) -> DefDict:
...
-    def all_enum(self, imports=True) -> dict[EnumDefinitionName, EnumDefinition]:
-        """
+    def all_enum(self, imports: bool = True) -> dict[EnumDefinitionName, EnumDefinition]:
+        """Retrieve all enums from the schema.
+
         :param imports: include imports closure
         :return: all enums in schema view
         """
-        return self._get_dict(ENUMS, imports)
+        return self.all_enums(imports)
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:553]: redirect to the non-deprecated function

File: linkml_runtime/utils/schemaview.py [hunk 8]
```diff
@@ -443,113 +500,118 @@ def _order_inheritance(self, elements: DefDict) -> DefDict:
...
         :return: all elements in schema view
         """
-        all_classes = self.all_classes(imports=imports)
-        all_slots = self.all_slots(imports=imports)
-        all_enums = self.all_enums(imports=imports)
-        all_types = self.all_types(imports=imports)
-        all_subsets = self.all_subsets(imports=imports)
-        # {**a,**b} syntax merges dictionary a and b into a single dictionary, removing duplicates.
-        return {**all_classes, **all_slots, **all_enums, **all_types, **all_subsets}
+        return self.all_elements(imports)
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:609]: redirect to non-deprecated function

File: linkml_runtime/utils/schemaview.py [hunk 9]
```diff
@@ -865,45 +941,63 @@ def class_ancestors(self, class_name: CLASS_NAME, imports=True, mixins=True, ref
         :param depth_first:
         :return: ancestor class names
         """
-        return _closure(lambda x: self.class_parents(x, imports=imports, mixins=mixins, is_a=is_a),
-                        class_name,
-                        reflexive=reflexive, depth_first=depth_first)
+        return _closure(
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:944]: reformatted

File: linkml_runtime/utils/schemaview.py [hunk 10]
```diff
@@ -1419,23 +1545,21 @@ def induced_slot(self, slot_name: SLOT_NAME, class_name: CLASS_NAME = None, impo
...
-                        # 1. if v2:
-                        # 2. if v2 is not None and
-                        #    (
-                        #      (isinstance(v2, (dict, list)) and v2) or
-                        #      (isinstance(v2, JsonObj) and as_dict(v2))
-                        #    )
-                        if not is_empty(v2):
-                            v = v2
-                            logger.debug(f'{v} takes precedence over {v2} for {induced_slot.name}.{metaslot_name}')
+                elif metaslot_name in COMBINE:
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:1548]: `else` + `if` => `elif`

File: linkml_runtime/utils/schemaview.py [hunk 11]
```diff
@@ -1745,47 +1853,53 @@ def usage_index(self) -> dict[ElementName, list[SchemaUsage]]:
...
 
     def add_subset(self, subset: SubsetDefinition) -> None:
-        """
+        """Add a subset to the schema.
+
         :param subset: subset to be added
         :return:
         """
-        self.schema.subsets[subset.name] = type
+        self.schema.subsets[subset.name] = subset
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:1897]: oops

File: linkml_runtime/utils/schemaview.py [hunk 12]
```diff
@@ -1745,47 +1853,53 @@ def usage_index(self) -> dict[ElementName, list[SchemaUsage]]:
...
-    def add_type(self, type: TypeDefinition) -> None:
-        """
+    def add_type(self, type_def: TypeDefinition) -> None:
+        """Add a type to the schema.
+
         :param type: type to be added
         :return:
         """
-        self.schema.types[type.name] = type
+        self.schema.types[type_def.name] = type_def
```
Comment (ialarmedalien) [inline linkml_runtime/utils/schemaview.py:1888]: `type` is a reserved word in python

File: tests/test_utils/test_schemaview.py [hunk 13]
```diff
@@ -895,10 +915,9 @@ def test_mergeimports():
     )
 
 
-def test_is_inlined():
-    schema_path = os.path.join(INPUT_DIR, "schemaview_is_inlined.yaml")
-    sv = SchemaView(schema_path)
-    cases = [
+@pytest.mark.parametrize(
```
Comment (ialarmedalien) [inline tests/test_utils/test_schemaview.py:918]: use parametrize instead of iterating through a list

File: tests/test_utils/test_schemaview.py [hunk 14]
```diff
@@ -44,26 +46,32 @@
...
 @pytest.fixture
-def schema_view_no_imports():
+def schema_view_no_imports() -> SchemaView:
+    """Fixture for SchemaView with a schema with no imports."""
     return SchemaView(SCHEMA_NO_IMPORTS)
 
 
 @pytest.fixture
-def view():
+def schema_view_with_imports() -> SchemaView:
```
Comment (ialarmedalien) [inline tests/test_utils/test_schemaview.py:55]: clearer name

File: tests/test_utils/test_schemaview.py [hunk 15]
```diff
@@ -44,26 +46,32 @@
...
     """Fixture for SchemaView with imports."""
     return SchemaView(SCHEMA_WITH_IMPORTS)
 
 
-def test_children_method(schema_view_no_imports):
+@pytest.fixture(scope="session")
+def schema_view_attributes() -> SchemaView:
```
Comment (ialarmedalien) [inline tests/test_utils/test_schemaview.py:61]: add a fixture for this since it's used in several tests

File: tests/test_utils/test_schemaview.py [hunk 16]
```diff
@@ -745,11 +758,9 @@ def test_slot_inheritance():
         view.slot_ancestors("s5")
 
 
-def test_attribute_inheritance():
-    """Tests attribute inheritance edge cases.
-    """
-    view = SchemaView(os.path.join(INPUT_DIR, "attribute_edge_cases.yaml"))
-    expected = [
+@pytest.mark.parametrize(
```
Comment (ialarmedalien) [inline tests/test_utils/test_schemaview.py:761]: parametric instead of iterating through a list

**Comment by Silvanoc** [comment]
@ialarmedalien do I understand it correctly? The conversations that you've opened in this PR are mostly justifying some of the changes, right? So a reviewer that holds those changes as meaningful, can simply comment "makes sense" or so and close those conversations, right?

**Comment by Silvanoc** [comment]
@ialarmedalien I'm trying to run "ruff" myself just to be sure that I don't look at changes directly applied by the tool and I'm failing. I'm getting an error message.

Doing so I've realized that we are not specifying `ruff` as a development requirement. The issue was with the version of `ruff` I have installed locally (a prehistoric v0.0.289). I've added `ruff` to the `dev` dependencies (which gives me v0.11.8), and now it runs.

I'll send a PR to your branch to add it.

**Comment by ialarmedalien** [comment]
> @ialarmedalien do I understand it correctly? The conversations that you've opened in this PR are mostly justifying some of the changes, right? So a reviewer that holds those changes as meaningful, can simply comment "makes sense" or so or "thumb-up" your comment and close those conversations, right?

The comments are to explain what I did, yes, to make life easier for the reviewer. There's no need to respond unless you particularly want to! 🙂

**Comment by ialarmedalien** [comment]
@Silvanoc re: ruff linting: I think it would be better if you made a direct PR to the linkml-runtime repo to update ruff and add it to the dev dependencies so that it gets the attention of the linkml-runtime maintainers.

The main reason I didn't add my config into this PR is that I have pretty much all the linters enabled locally, which I would not necessarily suggest for a repo like `linkml-runtime`.

This is my local config:

```
[tool.ruff]
line-length = 120
target-version = "py39"

[tool.ruff.lint]
select = [
    # core
    "F", # Pyflakes
    "E", # pycodestyle errors
    "W", # pycodestyle warnings
    "C90", # mccabe +
    "I", # isort
    "N", # pep8-naming
    "D", # pydocstyle
    "UP", # pyupgrade
    # extensions
    "YTT", # flake8-2020
    "ANN", # flake8-annotations
    "ASYNC", # flake8-async
    "S", # flake8-bandit
    "BLE", # flake8-blind-except
    "FBT", # flake8-boolean-trap
    "B", # flake8-bugbear
    "A", # flake8-builtins
    # "COM", # flake8-commas
    # "CPY", # flake8-copyright
    "C4", # flake8-comprehensions
    "DTZ", # flake8-datetimez
    "T10", # flake8-debugger
    # "DJ", # flake8-django
    "EM", # flake8-errmsg
    "EXE", # flake8-executable
    "FA", # flake8-future-annotations
    "ISC", # flake8-implicit-str-concat
    "ICN", # flake8-import-conventions
    "G", # flake8-logging-format
    "INP", # flake8-no-pep420
    "PIE", # flake8-pie
    "T20", # flake8-print
    "PYI", # flake8-pyi
    "PT", # flake8-pytest-style
    "Q", # flake8-quotes
    "RSE", # flake8-raise
    "RET", # flake8-return
    "SLF", # flake8-self
    "SLOT", # flake8-slots
    "SIM", # flake8-simplify
    "TID", # flake8-tidy-imports
    "TCH", # flake8-type-checking
    "INT", # flake8-gettext
    "ARG", # flake8-unused-arguments
    "PTH", # flake8-use-pathlib
    "TD", # flake8-todos
    "FIX", # flake8-fixme
    "ERA", # eradicate
    "PD", # pandas-vet
    "PGH", # pygrep-hooks
    "PL", # Pylint
    "TRY", # tryceratops
    "FLY", # flynt
    "NPY", # NumPy-specific rules
    "AIR", # Airflow
    "PERF", # Perflint
    "FURB", # refurb
    "LOG", # flake8-logging
    "RUF", # Ruff-specific rules
]

# Allow autofix for all enabled rules (when `--fix`) is provided.
fixable = ["ALL"]
unfixable = []

# D203: one-blank-line-before-class (conflicts with D211)
# D212: multi-line-summary-first-line (conflicts with D213)
# E203: whitespace before ',', ';', or ':'
# E501: line length
# ISC001: conflicts with Ruff's formatter
# W503: line break after binary operator
ignore = [
    "D203",
    "D213",
    "E203",
    "E501",
    "ISC001",
    "FBT001",
    "FBT002"
]

[tool.ruff.lint.per-file-ignores]
"tests/**/*.py" = ["S101"] # use of assert

[tool.ruff.lint.mccabe]
# Flag errors (`C901`) whenever the complexity level exceeds 15.
max-complexity = 15
```

I just ran `ruff check --fix`; I did not include the unsafe fixes.

I would love to format and lint auto-fix all the files in this repo, but there is an existing PR (#347) that does this and I didn't want to redo work that has already been done.

**Comment by ialarmedalien** [comment]
> IMO the whole linting, ruff, tox,... set-up in both `linkml` (especially there) and `linkml-runtime` is a mess. I would seriously consider cleaning it up...

I agree completely! I did some work on the `linkml-map` repo to introduce standardised ruff formatting and linting, which was fairly easy since the repo is low traffic. This repo would probably be the next good target before tackling the main `linkml` codebase.

**Comment by Silvanoc** [comment]
> @Silvanoc Apologies for unintentionally giving you a load of detective work to do in figuring out my ruff config! I would have included it as a comment on the PR if I had known.

No worries.

> If you look at the [last commit of the PR](https://github.com/linkml/linkml-runtime/pull/389/commits/c7e1c3a79f11c79ef748bbc5989db8373397714a), it contains the edits that I made manually or via VSCode's ruff plugin code fixing feature. Hopefully that will be a little easier to review...

I'm proposing [here](https://github.com/linkml/linkml-runtime/pull/347#issuecomment-2859179939) some changes to streamline linting and formatting. That way we have a clear definition of what we expect, we also enforce it and we enable developers to have it locally.

**Review by kevinschaper — APPROVED** [review]
thank you for the big cleanup! (and especially catching that add_subset bug!)

**Comment by Silvanoc** [comment]
Hmm, I'm not sure this PR was expected to get merged before #347. @ialarmedalien can you please confirm?
